pkgs/development/python-modules/hug/default.nix

```diff
@@ -6,6 +6,16 @@
 , mock
 , numpy
 }:
+let
+  falcon' = falcon.overridePythonAttrs (oldAttrs: rec {
```

Comment:
Hug might actually work with the current version of falcon. Please instead try to `substituteInPlace` the version constraint in hug's setup.py. Something like this:
```nix


  postPatch = ''
    substituteInPlace setup.py \
      --replace "==" ">="
  '';
```

Reply:
the tests fail
```
Sourcing python-remove-tests-dir-hook
Sourcing python-catch-conflicts-hook.sh
Sourcing python-remove-bin-bytecode-hook.sh
Sourcing setuptools-build-hook
Using setuptoolsBuildPhase
Using setuptoolsShellHook
Sourcing pip-install-hook
Using pipInstallPhase
Sourcing python-imports-check-hook.sh
Using pythonImportsCheckPhase
Sourcing python-namespaces-hook
Sourcing setuptools-check-hook
Using setuptoolsCheckPhase
Sourcing pytest-check-hook
Using pytestCheckPhase
Removing setuptoolsCheckPhase
@nix { "action": "setPhase", "phase": "unpackPhase" }
unpacking sources
unpacking source archive /nix/store/cqisgkpbmjg3vrhf3rq102jv5r3x014v-source
source root is source
setting SOURCE_DATE_EPOCH to timestamp 315619200 of file source/tox.ini
@nix { "action": "setPhase", "phase": "patchPhase" }
patching sources
@nix { "action": "setPhase", "phase": "configurePhase" }
configuring
no configure script, doing nothing
@nix { "action": "setPhase", "phase": "buildPhase" }
building
Executing setuptoolsBuildPhase
running bdist_wheel
The [wheel] section is deprecated. Use [bdist_wheel] instead.
running build
running build_py
creating build
creating build/lib
creating build/lib/hug
copying hug/validate.py -> build/lib/hug
copying hug/use.py -> build/lib/hug
copying hug/types.py -> build/lib/hug
copying hug/transform.py -> build/lib/hug
copying hug/this.py -> build/lib/hug
copying hug/test.py -> build/lib/hug
copying hug/store.py -> build/lib/hug
copying hug/routing.py -> build/lib/hug
copying hug/route.py -> build/lib/hug
copying hug/redirect.py -> build/lib/hug
copying hug/output_format.py -> build/lib/hug
copying hug/middleware.py -> build/lib/hug
copying hug/json_module.py -> build/lib/hug
copying hug/introspect.py -> build/lib/hug
copying hug/interface.py -> build/lib/hug
copying hug/input_format.py -> build/lib/hug
copying hug/format.py -> build/lib/hug
copying hug/exceptions.py -> build/lib/hug
copying hug/directives.py -> build/lib/hug
copying hug/development_runner.py -> build/lib/hug
copying hug/defaults.py -> build/lib/hug
copying hug/decorators.py -> build/lib/hug
copying hug/authentication.py -> build/lib/hug
copying hug/api.py -> build/lib/hug
copying hug/_version.py -> build/lib/hug
copying hug/_empty.py -> build/lib/hug
copying hug/__main__.py -> build/lib/hug
copying hug/__init__.py -> build/lib/hug
installing to build/bdist.linux-x86_64/wheel
running install
running install_lib
creating build/bdist.linux-x86_64
creating build/bdist.linux-x86_64/wheel
creating build/bdist.linux-x86_64/wheel/hug
copying build/lib/hug/__init__.py -> build/bdist.linux-x86_64/wheel/hug
copying build/lib/hug/__main__.py -> build/bdist.linux-x86_64/wheel/hug
copying build/lib/hug/_empty.py -> build/bdist.linux-x86_64/wheel/hug
copying build/lib/hug/_version.py -> build/bdist.linux-x86_64/wheel/hug
copying build/lib/hug/api.py -> build/bdist.linux-x86_64/wheel/hug
copying build/lib/hug/authentication.py -> build/bdist.linux-x86_64/wheel/hug
copying build/lib/hug/decorators.py -> build/bdist.linux-x86_64/wheel/hug
copying build/lib/hug/defaults.py -> build/bdist.linux-x86_64/wheel/hug
copying build/lib/hug/development_runner.py -> build/bdist.linux-x86_64/wheel/hug
copying build/lib/hug/directives.py -> build/bdist.linux-x86_64/wheel/hug
copying build/lib/hug/exceptions.py -> build/bdist.linux-x86_64/wheel/hug
copying build/lib/hug/format.py -> build/bdist.linux-x86_64/wheel/hug
copying build/lib/hug/input_format.py -> build/bdist.linux-x86_64/wheel/hug
copying build/lib/hug/interface.py -> build/bdist.linux-x86_64/wheel/hug
copying build/lib/hug/introspect.py -> build/bdist.linux-x86_64/wheel/hug
copying build/lib/hug/json_module.py -> build/bdist.linux-x86_64/wheel/hug
copying build/lib/hug/middleware.py -> build/bdist.linux-x86_64/wheel/hug
copying build/lib/hug/output_format.py -> build/bdist.linux-x86_64/wheel/hug
copying build/lib/hug/redirect.py -> build/bdist.linux-x86_64/wheel/hug
copying build/lib/hug/route.py -> build/bdist.linux-x86_64/wheel/hug
copying build/lib/hug/routing.py -> build/bdist.linux-x86_64/wheel/hug
copying build/lib/hug/store.py -> build/bdist.linux-x86_64/wheel/hug
copying build/lib/hug/test.py -> build/bdist.linux-x86_64/wheel/hug
copying build/lib/hug/this.py -> build/bdist.linux-x86_64/wheel/hug
copying build/lib/hug/transform.py -> build/bdist.linux-x86_64/wheel/hug
copying build/lib/hug/types.py -> build/bdist.linux-x86_64/wheel/hug
copying build/lib/hug/use.py -> build/bdist.linux-x86_64/wheel/hug
copying build/lib/hug/validate.py -> build/bdist.linux-x86_64/wheel/hug
running install_egg_info
running egg_info
creating hug.egg-info
writing hug.egg-info/PKG-INFO
writing dependency_links to hug.egg-info/dependency_links.txt
writing entry points to hug.egg-info/entry_points.txt
writing requirements to hug.egg-info/requires.txt
writing top-level names to hug.egg-info/top_level.txt
writing manifest file 'hug.egg-info/SOURCES.txt'
reading manifest file 'hug.egg-info/SOURCES.txt'
reading manifest template 'MANIFEST.in'
adding license file 'LICENSE'
writing manifest file 'hug.egg-info/SOURCES.txt'
Copying hug.egg-info to build/bdist.linux-x86_64/wheel/hug-2.6.0-py3.9.egg-info
running install_scripts
creating build/bdist.linux-x86_64/wheel/hug-2.6.0.dist-info/WHEEL
creating 'dist/hug-2.6.0-py2.py3-none-any.whl' and adding 'build/bdist.linux-x86_64/wheel' to it
adding 'hug/__init__.py'
adding 'hug/__main__.py'
adding 'hug/_empty.py'
adding 'hug/_version.py'
adding 'hug/api.py'
adding 'hug/authentication.py'
adding 'hug/decorators.py'
adding 'hug/defaults.py'
adding 'hug/development_runner.py'
adding 'hug/directives.py'
adding 'hug/exceptions.py'
adding 'hug/format.py'
adding 'hug/input_format.py'
adding 'hug/interface.py'
adding 'hug/introspect.py'
adding 'hug/json_module.py'
adding 'hug/middleware.py'
adding 'hug/output_format.py'
adding 'hug/redirect.py'
adding 'hug/route.py'
adding 'hug/routing.py'
adding 'hug/store.py'
adding 'hug/test.py'
adding 'hug/this.py'
adding 'hug/transform.py'
adding 'hug/types.py'
adding 'hug/use.py'
adding 'hug/validate.py'
adding 'hug-2.6.0.dist-info/LICENSE'
adding 'hug-2.6.0.dist-info/METADATA'
adding 'hug-2.6.0.dist-info/WHEEL'
adding 'hug-2.6.0.dist-info/entry_points.txt'
adding 'hug-2.6.0.dist-info/top_level.txt'
adding 'hug-2.6.0.dist-info/RECORD'
removing build/bdist.linux-x86_64/wheel
Finished executing setuptoolsBuildPhase
@nix { "action": "setPhase", "phase": "installPhase" }
installing
Executing pipInstallPhase
/build/source/dist /build/source
Processing ./hug-2.6.0-py2.py3-none-any.whl
Requirement already satisfied: falcon>=2.0.0 in /nix/store/hb5v1vplgamiddl5s4mzmgc1r0l4s14c-python3.9-falcon-3.0.1/lib/python3.9/site-packages (from hug==2.6.0) (3.0.1)
Requirement already satisfied: requests in /nix/store/4jhwmnr6f63xaxfrg5aw1kxhy0jc5sh3-python3.9-requests-2.26.0/lib/python3.9/site-packages (from hug==2.6.0) (2.26.0)
Requirement already satisfied: certifi>=2017.4.17 in /nix/store/91n7f6lqg9f9fm6dqc030nckr0x6hij4-python3.9-certifi-2021.05.30/lib/python3.9/site-packages (from requests->hug==2.6.0) (2021.5.30)
Requirement already satisfied: urllib3<1.27,>=1.21.1 in /nix/store/dmhc158ja8n7729gb38jwwh4z3p3kyfi-python3.9-urllib3-1.26.6/lib/python3.9/site-packages (from requests->hug==2.6.0) (1.26.6)
Requirement already satisfied: idna<4,>=2.5 in /nix/store/ppy0c4wg384wwgy8sj07d51yrfzk074l-python3.9-idna-3.2/lib/python3.9/site-packages (from requests->hug==2.6.0) (3.2)
Requirement already satisfied: charset-normalizer~=2.0.0 in /nix/store/j1jcsbcn4yvsf89j0lbh0wa8zway1ng7-python3.9-charset-normalizer-2.0.1/lib/python3.9/site-packages (from requests->hug==2.6.0) (2.0.1)
Installing collected packages: hug
Successfully installed hug-2.6.0
/build/source
Finished executing pipInstallPhase
@nix { "action": "setPhase", "phase": "fixupPhase" }
post-installation fixup
shrinking RPATHs of ELF executables and libraries in /nix/store/vzfa68yl2k8yz1h74j0y1dllybimvbj7-python3.9-hug-2.6.0
strip is /nix/store/a4mmjm3bblxwp8h53bcfx3dly80ib0ba-binutils-2.35.1/bin/strip
stripping (with command strip and flags -S) in /nix/store/vzfa68yl2k8yz1h74j0y1dllybimvbj7-python3.9-hug-2.6.0/lib  /nix/store/vzfa68yl2k8yz1h74j0y1dllybimvbj7-python3.9-hug-2.6.0/bin 
patching script interpreter paths in /nix/store/vzfa68yl2k8yz1h74j0y1dllybimvbj7-python3.9-hug-2.6.0
checking for references to /build/ in /nix/store/vzfa68yl2k8yz1h74j0y1dllybimvbj7-python3.9-hug-2.6.0...
Rewriting #!/nix/store/97w52ckcjnfiz89h3lh7zf1kysgfm2s8-python3-3.9.6/bin/python3.9 to #!/nix/store/97w52ckcjnfiz89h3lh7zf1kysgfm2s8-python3-3.9.6
wrapping `/nix/store/vzfa68yl2k8yz1h74j0y1dllybimvbj7-python3.9-hug-2.6.0/bin/hug'...
Executing pythonRemoveTestsDir
Finished executing pythonRemoveTestsDir
@nix { "action": "setPhase", "phase": "installCheckPhase" }
running install tests
no Makefile or custom installCheckPhase, doing nothing
@nix { "action": "setPhase", "phase": "pythonCatchConflictsPhase" }
pythonCatchConflictsPhase
@nix { "action": "setPhase", "phase": "pythonRemoveBinBytecodePhase" }
pythonRemoveBinBytecodePhase
@nix { "action": "setPhase", "phase": "pythonImportsCheckPhase" }
pythonImportsCheckPhase
Executing pythonImportsCheckPhase
@nix { "action": "setPhase", "phase": "pytestCheckPhase" }
pytestCheckPhase
Executing pytestCheckPhase
[1m============================= test session starts ==============================[0m
platform linux -- Python 3.9.6, pytest-6.2.5, py-1.10.0, pluggy-1.0.0
rootdir: /build/source, configfile: setup.cfg
[1mcollecting ... [0m[1m
collecting 445 items                                                           [0m[1m
collected 474 items / 24 deselected / 450 selected                             [0m

tests/test_api.py [32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[33m                                              [  2%][0m
tests/test_async.py [32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[33m                                                [  3%][0m
tests/test_authentication.py [32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[33m                                       [  4%][0m
tests/test_context_factory.py [32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[31mF[0m[32m.[0m[31m                            [  7%][0m
tests/test_coroutines.py [32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[31m                                           [  8%][0m
tests/test_decorators.py [32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[31mF[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[31mF[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[33ms[0m[31mF[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[31m [ 19%]
[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[31mF[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[31mF[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[31m                             [ 28%][0m
tests/test_directives.py [32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[31mF[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[31m                                  [ 32%][0m
tests/test_documentation.py [32m.[0m[32m.[0m[32m.[0m[32m.[0m[31m                                         [ 32%][0m
tests/test_exceptions.py [32m.[0m[31m                                               [ 33%][0m
tests/test_full_request.py [32m.[0m[31m                                             [ 33%][0m
tests/test_global_context.py [32m.[0m[31m                                           [ 33%][0m
tests/test_input_format.py [32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[31m                                         [ 34%][0m
tests/test_interface.py [32m.[0m[32m.[0m[32m.[0m[32m.[0m[31m                                             [ 35%][0m
tests/test_introspect.py [32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[31m                                         [ 37%][0m
tests/test_main.py [32m.[0m[31m                                                     [ 37%][0m
tests/test_middleware.py [32m.[0m[32m.[0m[32m.[0m[31m                                             [ 38%][0m
tests/test_output_format.py [32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[31m                            [ 41%][0m
tests/test_redirect.py [32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[31m                                            [ 43%][0m
tests/test_route.py [32m.[0m[32m.[0m[32m.[0m[31mF[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[31m                                     [ 46%][0m
tests/test_routing.py [32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[31m [ 57%]
[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[31m [ 73%]
[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[31m                 [ 86%][0m
tests/test_store.py [32m.[0m[31m                                                    [ 86%][0m
tests/test_test.py [32m.[0m[31m                                                     [ 86%][0m
tests/test_this.py [32m.[0m[31m                                                     [ 86%][0m
tests/test_types.py [32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[31m                   [ 94%][0m
tests/test_use.py [32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[31m                                 [ 99%][0m
tests/test_validate.py [32m.[0m[32m.[0m[32m.[0m[31m                                               [100%][0m

=================================== FAILURES ===================================
[31m[1m____________________ TestContextFactoryHTTP.test_exception _____________________[0m

self = <tests.test_context_factory.TestContextFactoryHTTP object at 0x7fff2df5bfa0>

    def test_exception(self):
        custom_context = dict(test="context")
    
        @hug.context_factory()
        def return_context(**kwargs):
            return custom_context
    
        @hug.delete_context()
        def delete_context(context, exception=None, errors=None, lacks_requirement=None):
            assert context == custom_context
            assert exception
            assert isinstance(exception, CustomException)
            assert not errors
            assert not lacks_requirement
            custom_context["launched_delete_context"] = True
    
        @hug.get("/exception_function")
        def exception_http_function():
            custom_context["launched_local_function"] = True
            raise CustomException()
    
        with pytest.raises(CustomException):
>           hug.test.get(module, "/exception_function")
[1m[31mE           Failed: DID NOT RAISE <class 'tests.test_context_factory.CustomException'>[0m

[1m[31mtests/test_context_factory.py[0m:518: Failed
----------------------------- Captured stderr call -----------------------------
2021-10-31 11:47:16 [FALCON] [ERROR] GET /exception_function => Traceback (most recent call last):
  File "/nix/store/hb5v1vplgamiddl5s4mzmgc1r0l4s14c-python3.9-falcon-3.0.1/lib/python3.9/site-packages/falcon/app.py", line 361, in __call__
    responder(req, resp, **params)
  File "/build/source/hug/interface.py", line 946, in __call__
    raise exception
  File "/build/source/hug/interface.py", line 917, in __call__
    self.call_function(input_parameters), context, request, response, **kwargs
  File "/build/source/hug/interface.py", line 839, in call_function
    return self.interface(**parameters)
  File "/build/source/hug/interface.py", line 129, in __call__
    return __hug_internal_self._function(*args, **kwargs)
  File "/build/source/tests/test_context_factory.py", line 515, in exception_http_function
    raise CustomException()
tests.test_context_factory.CustomException

[31m[1m____________________________ test_raise_on_invalid _____________________________[0m

    def test_raise_on_invalid():
        """Test to ensure hug correctly respects a request to allow validations errors to pass through as exceptions"""
    
        @hug.get(raise_on_invalid=True)
        def my_handler(argument_1: int):
            return True
    
        with pytest.raises(Exception):
>           hug.test.get(api, "my_handler", argument_1="hi")
[1m[31mE           Failed: DID NOT RAISE <class 'Exception'>[0m

[1m[31mtests/test_decorators.py[0m:244: Failed
----------------------------- Captured stderr call -----------------------------
2021-10-31 11:47:17 [FALCON] [ERROR] GET my_handler?argument_1=hi => Traceback (most recent call last):
  File "/build/source/hug/interface.py", line 362, in initialize_handler
    return handler(value, context=context)
TypeError: 'context' is an invalid keyword argument for int()

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/nix/store/hb5v1vplgamiddl5s4mzmgc1r0l4s14c-python3.9-falcon-3.0.1/lib/python3.9/site-packages/falcon/app.py", line 361, in __call__
    responder(req, resp, **params)
  File "/build/source/hug/interface.py", line 946, in __call__
    raise exception
  File "/build/source/hug/interface.py", line 911, in __call__
    errors = self.validate(input_parameters, context)
  File "/build/source/hug/interface.py", line 270, in validate
    input_parameters[key] = self.initialize_handler(
  File "/build/source/hug/interface.py", line 364, in initialize_handler
    return handler(value)
ValueError: invalid literal for int() with base 10: 'hi'

[31m[1m_______________________________ test_versioning ________________________________[0m

    def test_versioning():
        """Ensure that Hug correctly routes API functions based on version"""
    
        @hug.get("/echo")
        def echo(text):
            return "Not Implemented"
    
        @hug.get("/echo", versions=1)  # noqa
        def echo(text):
            return text
    
        @hug.get("/echo", versions=range(2, 4))  # noqa
        def echo(text):
            return "Echo: {text}".format(**locals())
    
        @hug.get("/echo", versions=7)  # noqa
        def echo(text, api_version):
            return api_version
    
        @hug.get("/echo", versions="8")  # noqa
        def echo(text, api_version):
            return api_version
    
        @hug.get("/echo", versions=False)  # noqa
        def echo(text):
            return "No Versions"
    
        with pytest.raises(ValueError):
    
            @hug.get("/echo", versions="eight")  # noqa
            def echo(text, api_version):
                return api_version
    
        assert hug.test.get(api, "v1/echo", text="hi").data == "hi"
        assert hug.test.get(api, "v2/echo", text="hi").data == "Echo: hi"
        assert hug.test.get(api, "v3/echo", text="hi").data == "Echo: hi"
        assert hug.test.get(api, "echo", text="hi", api_version=3).data == "Echo: hi"
        assert hug.test.get(api, "echo", text="hi", headers={"X-API-VERSION": "3"}).data == "Echo: hi"
        assert hug.test.get(api, "v4/echo", text="hi").data == "Not Implemented"
        assert hug.test.get(api, "v7/echo", text="hi").data == 7
        assert hug.test.get(api, "v8/echo", text="hi").data == 8
        assert hug.test.get(api, "echo", text="hi").data == "No Versions"
        assert (
            hug.test.get(api, "echo", text="hi", api_version=3, body={"api_vertion": 4}).data
            == "Echo: hi"
        )
    
        with pytest.raises(ValueError):
>           hug.test.get(api, "v4/echo", text="hi", api_version=3)
[1m[31mE           Failed: DID NOT RAISE <class 'ValueError'>[0m

[1m[31mtests/test_decorators.py[0m:438: Failed
----------------------------- Captured stderr call -----------------------------
2021-10-31 11:47:17 [FALCON] [ERROR] GET v4/echo?text=hi&api_version=3 => Traceback (most recent call last):
  File "/nix/store/hb5v1vplgamiddl5s4mzmgc1r0l4s14c-python3.9-falcon-3.0.1/lib/python3.9/site-packages/falcon/app.py", line 361, in __call__
    responder(req, resp, **params)
  File "/build/source/hug/api.py", line 352, in version_router
    request_version = self.determine_version(request, api_version)
  File "/build/source/hug/api.py", line 313, in determine_version
    raise ValueError("You are requesting conflicting versions")
ValueError: You are requesting conflicting versions

[31m[1m__________________________ test_marshmallow3_support ___________________________[0m

    @pytest.mark.skipif(MARSHMALLOW_MAJOR_VERSION != 3, reason="This test is for marshmallow 3 only")
    def test_marshmallow3_support():
        """Ensure that you can use Marshmallow style objects to control input and output validation and transformation"""
    
        class MarshmallowStyleObject(object):
            def dump(self, item):
                if item == "bad":
                    raise ValidationError("problems")
                return "Dump Success"
    
            def load(self, item):
                return "Load Success"
    
            def loads(self, item):
                return self.load(item)
    
        schema = MarshmallowStyleObject()
    
        @hug.get()
        def test_marshmallow_style() -> schema:
            return "world"
    
        assert hug.test.get(api, "test_marshmallow_style").data == "Dump Success"
        assert test_marshmallow_style() == "world"
    
        @hug.get()
        def test_marshmallow_style_error() -> schema:
            return "bad"
    
        with pytest.raises(InvalidTypeData):
>           hug.test.get(api, "test_marshmallow_style_error")
[1m[31mE           Failed: DID NOT RAISE <class 'hug.exceptions.InvalidTypeData'>[0m

[1m[31mtests/test_decorators.py[0m:697: Failed
----------------------------- Captured stderr call -----------------------------
2021-10-31 11:47:17 [FALCON] [ERROR] GET test_marshmallow_style_error => Traceback (most recent call last):
  File "/nix/store/hb5v1vplgamiddl5s4mzmgc1r0l4s14c-python3.9-falcon-3.0.1/lib/python3.9/site-packages/falcon/app.py", line 361, in __call__
    responder(req, resp, **params)
  File "/build/source/hug/interface.py", line 946, in __call__
    raise exception
  File "/build/source/hug/interface.py", line 916, in __call__
    self.render_content(
  File "/build/source/hug/interface.py", line 851, in render_content
    content = self.transform_data(content, request, response, context)
  File "/build/source/hug/interface.py", line 778, in transform_data
    return transform(data)
  File "/build/source/hug/types.py", line 732, in __call__
    raise InvalidTypeData(
hug.exceptions.InvalidTypeData: ('Invalid MarshmallowStyleObject passed in', ['problems'])

[31m[1m_______________________________ test_exceptions ________________________________[0m

    @pytest.mark.skipif(sys.platform == "win32", reason="Currently failing on Windows build")
    def test_exceptions():
        """Test to ensure hug's exception handling decorator works as expected"""
    
        @hug.get()
        def endpoint():
            raise ValueError("hi")
    
        with pytest.raises(ValueError):
>           hug.test.get(api, "endpoint")
[1m[31mE           Failed: DID NOT RAISE <class 'ValueError'>[0m

[1m[31mtests/test_decorators.py[0m:1646: Failed
----------------------------- Captured stderr call -----------------------------
2021-10-31 11:47:18 [FALCON] [ERROR] GET endpoint => Traceback (most recent call last):
  File "/nix/store/hb5v1vplgamiddl5s4mzmgc1r0l4s14c-python3.9-falcon-3.0.1/lib/python3.9/site-packages/falcon/app.py", line 361, in __call__
    responder(req, resp, **params)
  File "/build/source/hug/interface.py", line 946, in __call__
    raise exception
  File "/build/source/hug/interface.py", line 917, in __call__
    self.call_function(input_parameters), context, request, response, **kwargs
  File "/build/source/hug/interface.py", line 839, in call_function
    return self.interface(**parameters)
  File "/build/source/hug/interface.py", line 129, in __call__
    return __hug_internal_self._function(*args, **kwargs)
  File "/build/source/tests/test_decorators.py", line 1643, in endpoint
    raise ValueError("hi")
ValueError: hi

[31m[1m___________________________ test_exception_excludes ____________________________[0m

hug_api = <tests.fixtures.TestAPI object at 0x7fff2dc39680>

    def test_exception_excludes(hug_api):
        """Test to ensure it's possible to add excludes to exception routers"""
    
        class MyValueError(ValueError):
            pass
    
        class MySecondValueError(ValueError):
            pass
    
        @hug.exception(Exception, exclude=MySecondValueError, api=hug_api)
        def base_exception_handler(exception):
            return "base exception handler"
    
        @hug.exception(ValueError, exclude=(MyValueError, MySecondValueError), api=hug_api)
        def base_exception_handler(exception):
            return "special exception handler"
    
        @hug.get(api=hug_api)
        def my_handler():
            raise MyValueError()
    
        @hug.get(api=hug_api)
        def fall_through_handler():
            raise ValueError("reason")
    
        @hug.get(api=hug_api)
        def full_through_to_raise():
            raise MySecondValueError()
    
        assert hug.test.get(hug_api, "my_handler").data == "base exception handler"
        assert hug.test.get(hug_api, "fall_through_handler").data == "special exception handler"
        with pytest.raises(MySecondValueError):
>           assert hug.test.get(hug_api, "full_through_to_raise").data
[1m[31mE           Failed: DID NOT RAISE <class 'tests.test_decorators.test_exception_excludes.<locals>.MySecondValueError'>[0m

[1m[31mtests/test_decorators.py[0m:1866: Failed
----------------------------- Captured stderr call -----------------------------
2021-10-31 11:47:18 [FALCON] [ERROR] GET full_through_to_raise => Traceback (most recent call last):
  File "/nix/store/hb5v1vplgamiddl5s4mzmgc1r0l4s14c-python3.9-falcon-3.0.1/lib/python3.9/site-packages/falcon/app.py", line 361, in __call__
    responder(req, resp, **params)
  File "/build/source/hug/interface.py", line 940, in __call__
    raise exception
  File "/build/source/hug/interface.py", line 917, in __call__
    self.call_function(input_parameters), context, request, response, **kwargs
  File "/build/source/hug/interface.py", line 839, in call_function
    return self.interface(**parameters)
  File "/build/source/hug/interface.py", line 129, in __call__
    return __hug_internal_self._function(*args, **kwargs)
  File "/build/source/tests/test_decorators.py", line 1861, in full_through_to_raise
    raise MySecondValueError()
tests.test_decorators.test_exception_excludes.<locals>.MySecondValueError

[31m[1m_______________________________ test_current_api _______________________________[0m

    def test_current_api():
        """Ensure that it's possible to retrieve methods from the same version of the API"""
    
        @hug.get(versions=1)
        def first_method():
            return "Success"
    
        @hug.get(versions=1)
        def version_call_tester(hug_current_api):
            return hug_current_api.first_method()
    
        assert hug.test.get(api, "v1/version_call_tester").data == "Success"
    
        @hug.get()
        def second_method():
            return "Unversioned"
    
        @hug.get(versions=2)  # noqa
        def version_call_tester(hug_current_api):
            return hug_current_api.second_method()
    
        assert hug.test.get(api, "v2/version_call_tester").data == "Unversioned"
    
        @hug.get(versions=3)  # noqa
        def version_call_tester(hug_current_api):
            return hug_current_api.first_method()
    
        with pytest.raises(AttributeError):
>           hug.test.get(api, "v3/version_call_tester").data
[1m[31mE           Failed: DID NOT RAISE <class 'AttributeError'>[0m

[1m[31mtests/test_directives.py[0m:121: Failed
----------------------------- Captured stderr call -----------------------------
2021-10-31 11:47:18 [FALCON] [ERROR] GET v3/version_call_tester => Traceback (most recent call last):
  File "/nix/store/hb5v1vplgamiddl5s4mzmgc1r0l4s14c-python3.9-falcon-3.0.1/lib/python3.9/site-packages/falcon/app.py", line 361, in __call__
    responder(req, resp, **params)
  File "/build/source/hug/api.py", line 355, in version_router
    versions.get(request_version or False, versions.get(None, not_found))(
  File "/build/source/hug/interface.py", line 946, in __call__
    raise exception
  File "/build/source/hug/interface.py", line 917, in __call__
    self.call_function(input_parameters), context, request, response, **kwargs
  File "/build/source/hug/interface.py", line 839, in call_function
    return self.interface(**parameters)
  File "/build/source/hug/interface.py", line 129, in __call__
    return __hug_internal_self._function(*args, **kwargs)
  File "/build/source/tests/test_directives.py", line 118, in version_call_tester
    return hug_current_api.first_method()
  File "/build/source/hug/directives.py", line 126, in __getattr__
    raise AttributeError("API Function {0} not found".format(name))
AttributeError: API Function first_method not found

[31m[1m____________ test_routing_class_based_method_view_with_sub_routing _____________[0m

    def test_routing_class_based_method_view_with_sub_routing():
        """Test creating class based routers using method mappings, then overriding url on sub method"""
    
        @hug.object.http_methods()
        class EndPoint(object):
            def get(self):
                return "hi there!"
    
            @hug.object.urls("/home/")
            def post(self):
                return "bye"
    
        assert hug.test.get(api, "endpoint").data == "hi there!"
>       assert hug.test.post(api, "home").data == "bye"
[1m[31mE       AssertionError: assert {'errors': {'404 Not Found': None}} == 'bye'[0m
[1m[31mE        +  where {'errors': {'404 Not Found': None}} = <falcon.testing.srmock.StartResponseMock object at 0x7fff2d994040>.data[0m
[1m[31mE        +    where <falcon.testing.srmock.StartResponseMock object at 0x7fff2d994040> = functools.partial(<function call at 0x7fff2e777700>, 'POST')(<hug.api.API object at 0x7fff2d6524c0>, 'home')[0m
[1m[31mE        +      where functools.partial(<function call at 0x7fff2e777700>, 'POST') = <module 'hug.test' from '/build/source/hug/test.py'>.post[0m
[1m[31mE        +        where <module 'hug.test' from '/build/source/hug/test.py'> = hug.test[0m

[1m[31mtests/test_route.py[0m:103: AssertionError
[33m=============================== warnings summary ===============================[0m
hug/input_format.py:81
  /build/source/hug/input_format.py:81: SyntaxWarning: "is" with a literal. Did you mean "=="?
    if type(value) is list and len(value) is 1:

hug/api.py:41
  /build/source/hug/api.py:41: DeprecationWarning: invalid escape sequence \#
    INTRO = """

hug/use.py:185
  /build/source/hug/use.py:185: DeprecationWarning: invalid escape sequence \d
    status_code = int("".join(re.findall("\d+", response.status)))

hug/middleware.py:171
  /build/source/hug/middleware.py:171: DeprecationWarning: invalid escape sequence \d
    reqpath = re.sub("^(/v\d*/?)", "/", reqpath)

tests/test_use.py:106
  /build/source/tests/test_use.py:106: PytestUnknownMarkWarning: Unknown pytest.mark.extnetwork - is this a typo?  You can register custom marks to avoid this warning - for details, see https://docs.pytest.org/en/stable/mark.html
    @pytest.mark.extnetwork

tests/test_use.py:211
  /build/source/tests/test_use.py:211: PytestUnknownMarkWarning: Unknown pytest.mark.extnetwork - is this a typo?  You can register custom marks to avoid this warning - for details, see https://docs.pytest.org/en/stable/mark.html
    @pytest.mark.extnetwork

tests/test_async.py: 4 warnings
tests/test_authentication.py: 4 warnings
tests/test_context_factory.py: 5 warnings
tests/test_coroutines.py: 3 warnings
tests/test_decorators.py: 62 warnings
tests/test_directives.py: 10 warnings
tests/test_global_context.py: 1 warning
tests/test_interface.py: 1 warning
tests/test_middleware.py: 3 warnings
tests/test_output_format.py: 2 warnings
tests/test_route.py: 6 warnings
tests/test_routing.py: 1 warning
tests/test_types.py: 4 warnings
  /build/source/hug/api.py:361: DeprecatedWarning: Call to deprecated function __init__(...). API class may be removed in a future release, use falcon.App instead.
    falcon_api = self.falcon = falcon.API(middleware=self.middleware)

tests/test_authentication.py::test_basic_auth
tests/test_authentication.py::test_api_key
tests/test_authentication.py::test_token_auth
tests/test_authentication.py::test_missing_authenticator_docstring
  /build/source/hug/authentication.py:49: DeprecatedWarning: Calls with positional args are deprecated. Please specify them as keyword arguments instead.
    raise HTTPUnauthorized(

tests/test_authentication.py: 4 warnings
tests/test_context_factory.py: 1 warning
tests/test_decorators.py: 7 warnings
tests/test_directives.py: 1 warning
tests/test_output_format.py: 1 warning
tests/test_route.py: 1 warning
  /build/source/hug/api.py:411: DeprecatedWarning: Call to deprecated property body. Please use text instead.
    response.body = self.output_format(

tests/test_authentication.py::test_basic_auth
tests/test_authentication.py::test_api_key
tests/test_authentication.py::test_token_auth
  /build/source/hug/authentication.py:56: DeprecatedWarning: Calls with positional args are deprecated. Please specify them as keyword arguments instead.
    raise HTTPUnauthorized(

tests/test_authentication.py::test_basic_auth
  /build/source/hug/authentication.py:84: DeprecatedWarning: Calls with positional args are deprecated. Please specify them as keyword arguments instead.
    raise HTTPUnauthorized(

tests/test_authentication.py::test_basic_auth
  /build/source/hug/authentication.py:103: DeprecatedWarning: Calls with positional args are deprecated. Please specify them as keyword arguments instead.
    raise HTTPUnauthorized(

tests/test_coroutines.py::test_basic_call_coroutine
  /build/source/tests/test_coroutines.py:35: DeprecationWarning: "@coroutine" decorator is deprecated since Python 3.8, use "async def" instead
    def hello_world():

tests/test_coroutines.py::test_nested_basic_call_coroutine
  /build/source/tests/test_coroutines.py:46: DeprecationWarning: "@coroutine" decorator is deprecated since Python 3.8, use "async def" instead
    def hello_world():

tests/test_coroutines.py::test_nested_basic_call_coroutine
  /build/source/tests/test_coroutines.py:51: DeprecationWarning: "@coroutine" decorator is deprecated since Python 3.8, use "async def" instead
    def nested_hello_world():

tests/test_coroutines.py::test_basic_call_on_method_coroutine
  /build/source/tests/test_coroutines.py:63: DeprecationWarning: "@coroutine" decorator is deprecated since Python 3.8, use "async def" instead
    def hello_world(self=None):

tests/test_coroutines.py::test_basic_call_on_method_through_api_instance_coroutine
  /build/source/tests/test_coroutines.py:83: DeprecationWarning: "@coroutine" decorator is deprecated since Python 3.8, use "async def" instead
    def hello_world():

tests/test_coroutines.py::test_basic_call_on_method_registering_without_decorator_coroutine
  /build/source/tests/test_coroutines.py:98: DeprecationWarning: "@coroutine" decorator is deprecated since Python 3.8, use "async def" instead
    def hello_world_method(self):

tests/test_decorators.py::test_error_handling
  /build/source/tests/test_decorators.py:499: DeprecatedWarning: Calls with positional args are deprecated. Please specify them as keyword arguments instead.
    raise falcon.HTTPInternalServerError("Failed", "For Science!")

tests/test_decorators.py::test_startup
  /build/source/tests/test_decorators.py:1589: DeprecationWarning: "@coroutine" decorator is deprecated since Python 3.8, use "async def" instead
    def async_happens_on_startup(api):

tests/test_output_format.py::test_on_content_type
  /build/source/hug/output_format.py:344: DeprecatedWarning: Calls with positional args are deprecated. Please specify them as keyword arguments instead.
    raise falcon.HTTPNotAcceptable(error)

tests/test_output_format.py::test_accept
  /build/source/hug/output_format.py:393: DeprecatedWarning: Calls with positional args are deprecated. Please specify them as keyword arguments instead.
    raise falcon.HTTPNotAcceptable(error)

tests/test_output_format.py::test_accept_with_http_errors
  /build/source/tests/test_output_format.py:309: DeprecatedWarning: Calls with positional args are deprecated. Please specify them as keyword arguments instead.
    raise hug.HTTPInternalServerError("500 Internal Server Error", "This is an example")

tests/test_output_format.py::test_suffix
  /build/source/hug/output_format.py:425: DeprecatedWarning: Calls with positional args are deprecated. Please specify them as keyword arguments instead.
    raise falcon.HTTPNotAcceptable(error)

tests/test_output_format.py::test_prefix
  /build/source/hug/output_format.py:457: DeprecatedWarning: Calls with positional args are deprecated. Please specify them as keyword arguments instead.
    raise falcon.HTTPNotAcceptable(error)

tests/test_output_format.py::test_json_converter_numpy_types
  /build/source/tests/test_output_format.py:369: DeprecationWarning: `np.float` is a deprecated alias for the builtin `float`. To silence this warning, use `float` by itself. Doing this will not modify any behavior and is safe. If you specifically wanted the numpy scalar type, use `np.float64` here.
  Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations
    ex_np_float = numpy.float(0.5)

-- Docs: https://docs.pytest.org/en/stable/warnings.html
=========================== short test summary info ============================
FAILED tests/test_context_factory.py::TestContextFactoryHTTP::test_exception
FAILED tests/test_decorators.py::test_raise_on_invalid - Failed: DID NOT RAIS...
FAILED tests/test_decorators.py::test_versioning - Failed: DID NOT RAISE <cla...
FAILED tests/test_decorators.py::test_marshmallow3_support - Failed: DID NOT ...
FAILED tests/test_decorators.py::test_exceptions - Failed: DID NOT RAISE <cla...
FAILED tests/test_decorators.py::test_exception_excludes - Failed: DID NOT RA...
FAILED tests/test_directives.py::test_current_api - Failed: DID NOT RAISE <cl...
FAILED tests/test_route.py::test_routing_class_based_method_view_with_sub_routing
[31m==== [31m[1m8 failed[0m, [32m441 passed[0m, [33m1 skipped[0m, [33m24 deselected[0m, [33m150 warnings[0m[31m in 7.25s[0m[31m =====[0m
```